app/src/main/java/nodomain/freeyourgadget/gadgetbridge/contentprovider/HRContentProvider.java

```diff
@@ -0,0 +1,238 @@
+    private static final UriMatcher URI_MATCHER;
+
+    static {
+        URI_MATCHER = new UriMatcher(UriMatcher.NO_MATCH);
+        URI_MATCHER.addURI(HRContentProviderContract.AUTHORITY,
+                "devices", DEVICES_LIST);
+        URI_MATCHER.addURI(HRContentProviderContract.AUTHORITY,
+                "realtime", REALTIME);
+        URI_MATCHER.addURI(HRContentProviderContract.AUTHORITY,
+                "activity_start", ACTIVITY_START);
+        URI_MATCHER.addURI(HRContentProviderContract.AUTHORITY,
+                "activity_stop", ACTIVITY_STOP);
+    }
+
+    private ActivitySample buffered_sample = null;
+
+    // TODO: This is most of the time null...
+    private GBDevice mGBDevice = null;
+
+    private final BroadcastReceiver mReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            String action = intent.getAction();
+            //Log.i(HRContentProvider.class.getName(), "Received Event, aciton: " + action);
+
+            switch (action) {
+                case GBDevice.ACTION_DEVICE_CHANGED:
+                    mGBDevice = intent.getParcelableExtra(GBDevice.EXTRA_DEVICE);
+
+                    // Rationale: If device was not connected
+                    // it should show up here after beeing connected
+                    // If the user wanted to switch on realtime traffic, but we first needed to connect it
+                    // we do it here
+                    if (state == provider_state.ACTIVE) {
+                        GBApplication.deviceService().onEnableRealtimeSteps(true);
+                        GBApplication.deviceService().onEnableRealtimeHeartRateMeasurement(true);
+                    }
+                    break;
+                case DeviceService.ACTION_REALTIME_SAMPLES:
+                    ActivitySample tmp_sample = (ActivitySample) intent.getSerializableExtra(DeviceService.EXTRA_REALTIME_SAMPLE);
+                    if (tmp_sample.getHeartRate() == -1)
+                        break;
+
+                    buffered_sample = tmp_sample;
+                    // This notifies the observer
+                    getContext().
+                            getContentResolver().
+                            notifyChange(HRContentProviderContract.REALTIME_URI, null);
+                    break;
+                default:
+                    break;
+            }
+
+        }
+    };
+
+    @Override
+    public boolean onCreate() {
+        Log.i(HRContentProvider.class.getName(), "Creating...");
+        IntentFilter filterLocal = new IntentFilter();
+
+        filterLocal.addAction(GBDevice.ACTION_DEVICE_CHANGED);
+        filterLocal.addAction(DeviceService.ACTION_REALTIME_SAMPLES);
+
+        LocalBroadcastManager.getInstance(this.getContext()).registerReceiver(mReceiver, filterLocal);
+
+        //TODO Do i need only for testing or also in production?
+        this.getContext().registerReceiver(mReceiver, filterLocal);
```

Comment:
Gadgetbridge's own events are sent via LocalBroadcastManager, so this *should* not be necessary.

Reply:
Either way, for testing I need that hook. Maybe it is because I did not understand the testing framework completely. I'll have a look. I think there is a `GBApplication.isLocal()` call, so I can call it only if a testcase is running.